File: Python/Remove Nth Node From End of List.py
"""
Given a linked list, remove the nth node from the end of list and return its head.
For example,
   Given linked list: 1->2->3->4->5, and n = 2.

   After removing the second node from the end, the linked list becomes 1->2->3->5.
Note:
Given n will always be valid.
Try to do this in one pass.
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @return a ListNode
    def removeNthFromEnd(self, head, n):
        if head is None:
            return head
        count = n
        p1 = head
        p2 = head
        while n > 0:
            p2 = p2.next
            n -=1
        if p2 is None:
            return head.next
        prev = head
        while p2:
            prev = p1
            p1 = p1.next
            p2 = p2.next
        prev.next = p1.next
        return head

File: Java/Remove Nth Node From End of List.java
/*
Given a linked list, remove the nth node from the end of list and return its head.

For example,

   Given linked list: 1->2->3->4->5, and n = 2.

   After removing the second node from the end, the linked list becomes 1->2->3->5.
Note:
Given n will always be valid.
Try to do this in one pass.
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if(head == null) return head;
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode prev = dummy;
        ListNode cur = dummy;
        for(int i = 0 ; i < n ; i ++){
            prev = prev.next;
        }
        
        while(prev.next != null){
            prev = prev.next;
            cur = cur.next;
        }
        
        cur.next = cur.next.next;
        
        return dummy.next;
    }
}